src/ralph/api/auth/oidc.py

```diff
@@ -0,0 +1,116 @@
+"""OpenID Connect authentication tool for the Ralph API."""
+
+import logging
+from functools import lru_cache
+from typing import Union
+
+import requests
+from fastapi import Depends, HTTPException, status
+from fastapi.security import OpenIdConnect
+from jose import ExpiredSignatureError, JWTError, jwt
+from jose.exceptions import JWTClaimsError
+from pydantic import BaseModel, Extra
+
+from ralph.conf import settings
+
+oauth2_scheme = OpenIdConnect(
+    openIdConnectUrl=f"{settings.AUTH_PROVIDER_URL}"
+    "/.well-known/openid-configuration",
+    auto_error=False,
+)
+
+# API auth logger
+logger = logging.getLogger(__name__)
+
+
+class IDToken(BaseModel):
+    """Pydantic model representing an OpenID Connect ID Token.
+
+    ID Tokens are polymorphic and may have many attributes not defined in the
+    spec thus this model accepts all addition fields. Only required fields are
+    listed in the attributes section of this docstring or enforced by pydantic.
+
+    Attributes:
+        iss (str): Issuer Identifier for the Issuer of the response.
+        sub (str): Subject Identifier.
+        aud (str): Audience(s) that this ID Token is intended for.
+        exp (str): Expiration time on or after which the ID Token MUST NOT be
+                   accepted for processing.
+        iat (iat): Time at which the JWT was issued.
+    """
+
+    iss: str
+    sub: str
+    aud: str
+    exp: int
+    iat: int
+
+    class Config:  # pylint: disable=missing-class-docstring # noqa: D106
+        extra = Extra.allow
+
+
+def requests_get(url: str) -> dict:
+    """Send a get request to the url.
+
+    Args:
+        url (str): URL endpoint  to send the get request to.
+
+    Return:
+        dict
+
+    Raises:
+        HTTPException HTTP_401_UNAUTHORIZED
+    """
+    try:
+        response = requests.get(url, timeout=5)
+        response.raise_for_status()
+        return response.json()
+    except requests.exceptions.RequestException as exc:
+        raise HTTPException(
+            status_code=status.HTTP_401_UNAUTHORIZED,
+            detail="Could not validate credentials",
+            headers={"WWW-Authenticate": "Bearer"},
+        ) from exc
+
+
+@lru_cache()
+def discover_provider(base_url: str) -> dict:
+    """Discover the authentication server (or OpenId Provider) configuration."""
+    return requests_get(f"{base_url}/.well-known/openid-configuration")
```

Comment:
I think the `requests_get` helper is not required. As it's only used twice, you can duplicate this snippet and add specific logging messages (`logger.ERROR`) so that it's easier to debug OIDC configuration for operators.